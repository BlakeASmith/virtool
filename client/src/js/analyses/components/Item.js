import React from "react";
import CX from "classnames";
import { get } from "lodash-es";
import { connect } from "react-redux";
import { LinkContainer } from "react-router-bootstrap";
import { Row, Col, Label } from "react-bootstrap";

import { getTaskDisplayName } from "../../utils/utils";
import { Icon, Loader, RelativeTime } from "../../base";
import { removeAnalysis } from "../actions";
import { getCanModify } from "../../samples/selectors";

export const RightIcon = ({ canModify, onRemove, ready }) => {
    if (ready) {
        if (canModify) {
            return <Icon name="trash" bsStyle="danger" onClick={onRemove} style={{ fontSize: "17px" }} pullRight />;
        }

        return null;
    }

    return (
        <div className="pull-right">
            <Loader size="14px" color="#3c8786" />
        </div>
    );
};

export const AnalysisItem = props => {
    const itemClass = CX("list-group-item spaced", { hoverable: props.ready });

    const reference = props.placeholder ? null : (
        <span>
            {props.reference.name}
            <Label style={{ marginLeft: "5px" }}>{props.index.version}</Label>
        </span>
    );

    const content = (
        <div className={itemClass} style={{ color: "#555" }}>
            <Row>
                <Col xs={4} sm={4} md={4}>
                    <strong>{getTaskDisplayName(props.algorithm)}</strong>
                </Col>
                <Col xs={5} sm={4} md={4}>
                    Started <RelativeTime time={props.created_at} />
                    {props.placeholder ? null : ` by ${props.user.id}`}
                </Col>
                <Col xs={2}>{reference}</Col>
                <Col xsHidden sm={2} md={2}>
                    <RightIcon {...props} />
                </Col>
            </Row>
        </div>
    );

    if (props.placeholder) {
        return content;
    }

    return <LinkContainer to={`/samples/${props.sampleId}/analyses/${props.id}`}>{content}</LinkContainer>;
};

const mapStateToProps = state => ({
    sampleId: get(state.samples.detail, "id"),
    canModify: getCanModify(state)
});

const mapDispatchToProps = (dispatch, ownProps) => ({
    onRemove: () => {
        dispatch(removeAnalysis(ownProps.id));
    }
});

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(AnalysisItem);
