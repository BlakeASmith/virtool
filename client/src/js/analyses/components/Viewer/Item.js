import styled from "styled-components";
import { Box } from "../../../base";

export const AnalysisViewerItem = styled(Box)`
    border-bottom: none;
    border-left: none;
    border-radius: 0;
    margin: 0;
    ${props => (props.selected ? `box-shadow: inset 3px 0 0 ${props.theme.color.primary};` : "")}
`;
