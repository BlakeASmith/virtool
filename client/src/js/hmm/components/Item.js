import React from "react";
import PropTypes from "prop-types";
import { LinkContainer } from "react-router-bootstrap";
import { keys, map, reject } from "lodash-es";
import { Col, Row } from "react-bootstrap";

import { Label, ListGroupItem } from "../../base";

export default function HMMItem({ cluster, families, id, names }) {
    const filteredFamilies = reject(keys(families), family => family === "None");

    const labelComponents = map(filteredFamilies.slice(0, 3), (family, i) => (
        <Label key={i} spaced>
            {family}
        </Label>
    ));

    return (
        <LinkContainer to={`/hmm/${id}`}>
            <ListGroupItem className="spaced">
                <Row>
                    <Col xs={2}>
                        <strong>{cluster}</strong>
                    </Col>
                    <Col xs={5}>{names[0]}</Col>
                    <Col xs={5}>
                        <div className="pull-right">
                            {labelComponents} {filteredFamilies.length > 3 ? "..." : null}
                        </div>
                    </Col>
                </Row>
            </ListGroupItem>
        </LinkContainer>
    );
}

HMMItem.propTypes = {
    cluster: PropTypes.number,
    families: PropTypes.object,
    id: PropTypes.string,
    names: PropTypes.array
};
