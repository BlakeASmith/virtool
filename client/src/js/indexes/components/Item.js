import React from "react";
import { connect } from "react-redux";
import { LinkContainer } from "react-router-bootstrap";
import { Row, Col, Label } from "react-bootstrap";
import { Icon, RelativeTime, ListGroupItem, Loader } from "../../base";
import { activeIndexIdSelector } from "../selectors";

export const IndexEntry = ({ activeId, document, refId }) => {
    let activeIcon;

    // Decide what icon/text should be shown at the right end of the index document. If the index is building a
    // spinner with "Building" is shown, if the index is the active index a green check is shown. Otherwise, no
    // content is shown at the right.
    if (document.id === activeId && document.ready) {
        activeIcon = (
            <span className="pull-right">
                <Icon name="check" bsStyle="success" /> <strong>Active</strong>
            </span>
        );
    }

    if (!document.ready) {
        activeIcon = (
            <div className="pull-right">
                <Loader size="14px" color="#3c8786" style={{ display: "inline" }} />
                <strong> Building</strong>
            </div>
        );
    }

    // The description of
    let changeDescription;

    if (document.change_count !== null) {
        // Text to show if no changes occurred since the last index build. Technically, should never be shown
        // because the rebuild button is not shown if no changes have been made.
        changeDescription = "No changes";

        // This should always test true in practice. Shows the number of changes and the number of OTUs
        // affected.
        if (document.change_count > 0) {
            changeDescription = (
                <span>
                    {document.change_count} {" changes made in "}
                    {document.modified_otu_count} {document.modified_otu_count === 1 ? "OTU" : "OTUs"}
                </span>
            );
        }
    }

    return (
        <LinkContainer to={`/refs/${refId}/indexes/${document.id}`} className="spaced">
            <ListGroupItem>
                <Row>
                    <Col xs={3}>
                        <strong>Version {document.version}</strong>
                    </Col>
                    <Col xs={3}>
                        Created <RelativeTime time={document.created_at} />
                    </Col>
                    <Col md={4} xsHidden smHidden>
                        {changeDescription}
                    </Col>
                    <Col xs={6} md={2}>
                        {activeIcon}
                    </Col>
                </Row>
            </ListGroupItem>
        </LinkContainer>
    );
};

const mapStateToProps = (state, ownProps) => ({
    document: state.indexes.documents[ownProps.index],
    activeId: activeIndexIdSelector(state),
    refId: state.references.detail.id
});

export default connect(mapStateToProps)(IndexEntry);
