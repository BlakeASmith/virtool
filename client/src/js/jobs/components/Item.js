import React from "react";
import PropTypes from "prop-types";
import { capitalize } from "lodash-es";
import { Row, Col } from "react-bootstrap";
import { connect } from "react-redux";
import { push } from "connected-react-router";
import { RelativeTime, ProgressBar, Flex, FlexItem } from "../../base";
import { getTaskDisplayName } from "../../utils/utils";
import { cancelJob, removeJob } from "../actions";
import { JobActionIcon, JobStatusIcon } from "./Icons";

export class JobItem extends React.Component {
    static propTypes = {
        id: PropTypes.string.isRequired,
        task: PropTypes.string.isRequired,
        state: PropTypes.string.isRequired,
        progress: PropTypes.number.isRequired,
        created_at: PropTypes.string.isRequired,
        user: PropTypes.object.isRequired,
        onNavigate: PropTypes.func,
        onCancel: PropTypes.func,
        onRemove: PropTypes.func,
        canRemove: PropTypes.bool,
        canCancel: PropTypes.bool
    };

    handleCancel = e => {
        e.stopPropagation();
        this.props.onCancel(this.props.id);
    };

    handleNavigate = e => {
        e.stopPropagation();
        this.props.onNavigate(this.props.id);
    };

    handleRemove = e => {
        e.stopPropagation();
        this.props.onRemove(this.props.id);
    };

    render() {
        let progressStyle = "success";

        const progressValue = this.props.progress * 100;

        if (this.props.state === "running") {
            progressStyle = "warning";
        }

        if (this.props.state === "error" || this.props.state === "cancelled") {
            progressStyle = "danger";
        }

        // Create the option components for the selected fields.
        return (
            <div
                className="list-group-item hoverable spaced job"
                onClick={this.handleNavigate}
                style={{ color: "#555" }}
            >
                <ProgressBar now={progressValue} bsStyle={progressStyle} affixed />

                <Row>
                    <Col xs={6} sm={4} md={4}>
                        <strong>{getTaskDisplayName(this.props.task)}</strong>
                    </Col>
                    <Col xs={5} sm={5} md={5}>
                        Started <RelativeTime time={this.props.created_at} /> by {this.props.user.id}
                    </Col>
                    <Col xsHidden sm={2} md={2}>
                        <Flex justifyContent="flex-end">
                            <FlexItem>
                                <JobStatusIcon state={this.props.state} />
                            </FlexItem>
                            <FlexItem pad>
                                <strong>{capitalize(this.props.state)}</strong>
                            </FlexItem>
                        </Flex>
                    </Col>
                    <Col xs={1} sm={1} md={1}>
                        <JobActionIcon
                            state={this.props.state}
                            canCancel={this.props.canCancel}
                            canRemove={this.props.canRemove}
                            onCancel={this.onCancel}
                            onRemove={this.onRemove}
                        />
                    </Col>
                </Row>
            </div>
        );
    }
}

export const mapDispatchToProps = dispatch => ({
    onCancel: jobId => {
        dispatch(cancelJob(jobId));
    },

    onNavigate: jobId => {
        dispatch(push(`/jobs/${jobId}`));
    },

    onRemove: jobId => {
        dispatch(removeJob(jobId));
    }
});

export default connect(
    null,
    mapDispatchToProps
)(JobItem);
