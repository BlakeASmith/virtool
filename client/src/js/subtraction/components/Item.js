import React from "react";
import { connect } from "react-redux";
import { LinkContainer } from "react-router-bootstrap";
import { Col, Row } from "react-bootstrap";
import { Flex, FlexItem, Icon, ListGroupItem, Loader } from "../../base";

export const SubtractionItemIcon = ({ ready }) => {
    if (ready) {
        return <Icon name="check" bsStyle="success" />;
    }

    return <Loader size="14px" color="#3c8786" />;
};

export const SubtractionItem = ({ id, ready }) => (
    <LinkContainer key={id} className="spaced" to={`/subtraction/${id}`}>
        <ListGroupItem>
            <Row>
                <Col xs={8} md={8}>
                    <strong>{id}</strong>
                </Col>
                <Col xs={4} md={4}>
                    <Flex alignItems="center" className="pull-right">
                        <SubtractionItemIcon ready={ready} />
                        <FlexItem pad>
                            <strong>{ready ? "Ready" : "Importing"}</strong>
                        </FlexItem>
                    </Flex>
                </Col>
            </Row>
        </ListGroupItem>
    </LinkContainer>
);

export const mapStateToProps = (state, props) => {
    const { id, ready } = state.subtraction.documents[props.index];
    return {
        id,
        ready
    };
};

export default connect(mapStateToProps)(SubtractionItem);
