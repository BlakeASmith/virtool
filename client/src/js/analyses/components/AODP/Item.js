import React, { useCallback } from "react";
import { connect } from "react-redux";

import { setActiveHitId } from "../../actions";
import { getActiveHit, getMatches } from "../../selectors";
import { AnalysisViewerItem } from "../Viewer/Item";

export const AODPItem = ({ active, name, id, style, onSetActiveId }) => {
    const handleClick = useCallback(() => onSetActiveId(id), [id]);

    return (
        <AnalysisViewerItem onClick={handleClick} style={style} selected={active}>
            <strong>{name}</strong>
        </AnalysisViewerItem>
    );
};

const mapStateToProps = (state, ownProps) => {
    const activeId = getActiveHit(state).id;
    const matches = getMatches(state);
    const match = matches[ownProps.index];

    const { id, name } = match;
    return { id, name, active: activeId === id };
};

const mapDispatchToProps = dispatch => ({
    onSetActiveId: id => {
        dispatch(setActiveHitId(id));
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(AODPItem);
