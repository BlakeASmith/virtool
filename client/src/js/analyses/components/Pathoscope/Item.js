import React, { useCallback } from "react";
import { connect } from "react-redux";
import styled from "styled-components";
import { SlashList } from "../../../base";
import { setActiveHitId } from "../../actions";
import { getActiveHit, getMatches } from "../../selectors";
import { AnalysisViewerItem } from "../Viewer/Item";
import { StaticOTUCoverage } from "./OTUCoverage";

const PathoscopeItemTitle = styled.div`
    display: block;
    overflow: hidden;
    font-weight: bold;
    text-overflow: ellipsis;
    white-space: nowrap;
`;

const PathoscopeItemSubtitle = styled.small`
    align-items: center;
    display: flex;
    justify-content: space-between;
`;

const StyledPathoscopeItem = styled(AnalysisViewerItem)`
    height: 142px;
`;

export const PathoscopeItem = ({ active, hit, style, onSetActiveId }) => {
    const { abbreviation, coverage, depth, id, filled, name, pi } = hit;

    const handleClick = useCallback(() => onSetActiveId(id), [id]);

    return (
        <StyledPathoscopeItem selected={active} style={style} onClick={handleClick}>
            <PathoscopeItemTitle>{name}</PathoscopeItemTitle>
            <PathoscopeItemSubtitle>
                <span>{abbreviation || "No Abbreviation"}</span>
                <SlashList>
                    <li className="text-danger">{depth}</li>
                    <li className="text-success">{pi.toFixed(3)}</li>
                    <li className="text-primary">{coverage.toFixed(3)}</li>
                </SlashList>
            </PathoscopeItemSubtitle>
            <StaticOTUCoverage id={id} filled={filled} />
        </StyledPathoscopeItem>
    );
};

const mapStateToProps = (state, ownProps) => {
    const activeId = getActiveHit(state).id;
    const hit = getMatches(state)[ownProps.index];

    return { hit, active: activeId === hit.id };
};

const mapDispatchToProps = dispatch => ({
    onSetActiveId: id => {
        dispatch(setActiveHitId(id));
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(PathoscopeItem);
