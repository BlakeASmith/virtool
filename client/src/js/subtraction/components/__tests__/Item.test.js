import { SubtractionItem, SubtractionItemIcon, mapStateToProps } from "../Item";

describe("<SubtractionItemIcon />", () => {
    it.each([true, false])("should render when [ready=%p]", ready => {
        const wrapper = shallow(<SubtractionItemIcon ready={ready} />);
        expect(wrapper).toMatchSnapshot();
    });
});

describe("<SubtractionItem />", () => {
    let props;

    beforeEach(() => {
        props = {
            id: "Foo Bar",
            ready: true
        };
    });

    it.each([true, false])("should render when [ready=%p]", ready => {
        props.ready = ready;
        const wrapper = shallow(<SubtractionItem {...props} />);
        expect(wrapper).toMatchSnapshot();
    });
});

describe("mapStateToProps()", () => {
    it("should return props", () => {
        const state = {
            subtraction: {
                documents: [
                    { id: "foo", ready: true },
                    { id: "bar", ready: true },
                    { id: "baz", ready: true }
                ]
            }
        };
        const props = mapStateToProps(state, { index: 1 });
        expect(props).toEqual({
            id: "bar",
            ready: true
        });
    });
});
