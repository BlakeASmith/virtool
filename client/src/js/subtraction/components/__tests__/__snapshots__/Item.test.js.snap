// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SubtractionItem /> should render when [ready=false] 1`] = `
<withRouter(LinkContainer)
  className="spaced"
  key="Foo Bar"
  to="/subtraction/Foo Bar"
>
  <ListGroupItem
    allowFocus={false}
  >
    <Row
      bsClass="row"
      componentClass="div"
    >
      <Col
        bsClass="col"
        componentClass="div"
        md={8}
        xs={8}
      >
        <strong>
          Foo Bar
        </strong>
      </Col>
      <Col
        bsClass="col"
        componentClass="div"
        md={4}
        xs={4}
      >
        <Flex
          alignContent="stretch"
          alignItems="center"
          className="pull-right"
          direction="row"
          justifyContent="flex-start"
          wrap="nowrap"
        >
          <SubtractionItemIcon
            ready={false}
          />
          <FlexItem
            alignSelf={null}
            basis="auto"
            grow={0}
            pad={true}
            shrink={1}
          >
            <strong>
              Importing
            </strong>
          </FlexItem>
        </Flex>
      </Col>
    </Row>
  </ListGroupItem>
</withRouter(LinkContainer)>
`;

exports[`<SubtractionItem /> should render when [ready=true] 1`] = `
<withRouter(LinkContainer)
  className="spaced"
  key="Foo Bar"
  to="/subtraction/Foo Bar"
>
  <ListGroupItem
    allowFocus={false}
  >
    <Row
      bsClass="row"
      componentClass="div"
    >
      <Col
        bsClass="col"
        componentClass="div"
        md={8}
        xs={8}
      >
        <strong>
          Foo Bar
        </strong>
      </Col>
      <Col
        bsClass="col"
        componentClass="div"
        md={4}
        xs={4}
      >
        <Flex
          alignContent="stretch"
          alignItems="center"
          className="pull-right"
          direction="row"
          justifyContent="flex-start"
          wrap="nowrap"
        >
          <SubtractionItemIcon
            ready={true}
          />
          <FlexItem
            alignSelf={null}
            basis="auto"
            grow={0}
            pad={true}
            shrink={1}
          >
            <strong>
              Ready
            </strong>
          </FlexItem>
        </Flex>
      </Col>
    </Row>
  </ListGroupItem>
</withRouter(LinkContainer)>
`;

exports[`<SubtractionItemIcon /> should render when [ready=false] 1`] = `
<Loader
  color="#3c8786"
  size="14px"
/>
`;

exports[`<SubtractionItemIcon /> should render when [ready=true] 1`] = `
<Icon
  bsStyle="success"
  faStyle="fas"
  fixedWidth={false}
  name="check"
  pullRight={false}
/>
`;
